typing/typedecl.ml

```diff
@@ -1393,9 +1393,34 @@ let native_repr_of_type env kind ty =
   | _ ->
     None
 
+let warns_if_has_native_repr_attributes core_type =
+  let open Ast_mapper in
+  let this_mapper =
+    { default_mapper with typ = fun mapper core_type ->
+      begin
+        match
+          get_native_repr_attribute core_type.ptyp_attributes ~global_repr:None
+        with
+        | Native_repr_attr_present kind ->
+          Location.prerr_warning core_type.ptyp_loc
+            (Warnings.Ignored_attribute_in_external_declaration
+              (Printf.sprintf
+                 "'%s' should be attached to a direct argument or result of \
+                  the primitive, it should not occur deeply into its type"
+              (match kind with
+                  Unboxed -> "@unboxed"
+                | Untagged -> "@untagged")))
+        | Native_repr_attr_absent -> ()
+      end;
+      default_mapper.typ mapper core_type }
+  in
+  ignore (this_mapper.typ this_mapper core_type)
+
 let make_native_repr env core_type ty ~global_repr =
   match get_native_repr_attribute core_type.ptyp_attributes ~global_repr with
-  | Native_repr_attr_absent -> Same_as_ocaml_repr
+  | Native_repr_attr_absent ->
+    warns_if_has_native_repr_attributes core_type;
```

Comment:
We should do the check in both cases. We just need to make sure we don't warn for attributes at the toplevel:

```ocaml
ignore (default_mapper.typ this_mapper core_type : core_type)
```